pkg/vfs/couchdb_indexer.go

```diff
@@ -136,36 +137,82 @@ func (c *couchdbIndexer) CreateNamedDirDoc(doc *DirDoc) error {
 func (c *couchdbIndexer) UpdateDirDoc(olddoc, newdoc *DirDoc) error {
 	newdoc.SetID(olddoc.ID())
 	newdoc.SetRev(olddoc.Rev())
+
+	oldTrashed := strings.HasPrefix(olddoc.Fullpath, TrashDirName)
+	newTrashed := strings.HasPrefix(newdoc.Fullpath, TrashDirName)
+
+	isRestored := oldTrashed && !newTrashed
+	isTrashed := !oldTrashed && newTrashed
+
+	if isTrashed {
+		if err := c.setTrashedForFilesInsideDir(olddoc, true); err != nil {
+			return err
+		}
+	}
+
 	if newdoc.Fullpath != olddoc.Fullpath {
 		if err := c.moveDir(olddoc.Fullpath, newdoc.Fullpath); err != nil {
 			return err
 		}
 	}
-	return couchdb.UpdateDoc(c.db, newdoc)
+
+	if err := couchdb.UpdateDoc(c.db, newdoc); err != nil {
+		return err
+	}
+
+	if isRestored {
+		if err := c.setTrashedForFilesInsideDir(newdoc, false); err != nil {
+			return err
+		}
+	}
+
+	return nil
 }
 
 func (c *couchdbIndexer) DeleteDirDoc(doc *DirDoc) error {
 	return couchdb.DeleteDoc(c.db, doc)
 }
 
 func (c *couchdbIndexer) moveDir(oldpath, newpath string) error {
+	var docs []interface{}
 	var children []*DirDoc
-	sel := mango.StartWith("path", oldpath+"/")
-	req := &couchdb.FindRequest{
-		UseIndex: "dir-by-path",
-		Selector: sel,
-	}
-	err := couchdb.FindDocs(c.db, consts.Files, req, &children)
-	if err != nil || len(children) == 0 {
-		return err
-	}
 
-	couchdocs := make([]interface{}, len(children))
-	for i, child := range children {
-		child.Fullpath = path.Join(newpath, child.Fullpath[len(oldpath)+1:])
-		couchdocs[i] = child
+	skip := 0
+	for {
+		limit := 256
+		sel := mango.StartWith("path", oldpath+"/")
+		req := &couchdb.FindRequest{
+			UseIndex: "dir-by-path",
+			Selector: sel,
+			Skip:     skip,
+			Limit:    limit,
+		}
+		err := couchdb.FindDocs(c.db, consts.Files, req, &children)
+		if err != nil {
+			return err
+		}
+		if len(children) == 0 {
+			break
+		}
+		if cap(docs) < len(children) {
+			docs = make([]interface{}, 0, len(children))
+		}
+		for _, child := range children {
+			child.Fullpath = path.Join(newpath, child.Fullpath[len(oldpath)+1:])
+			docs = append(docs, child)
+		}
+		if err = couchdb.BulkUpdateDocs(c.db, consts.Files, docs); err != nil {
```

Comment:
I'm not sure I follow you here: why do you copy documents from `children` to `docs`, and not use `children` directly?